pypsa/networkclustering.py

```diff
@@ -513,6 +514,85 @@ def rectangular_grid_clustering(network, divisions):
     busmap = busmap_by_rectangular_grid(network.buses, divisions)
     return get_clustering_from_busmap(network, busmap)
 
+################
+# Hierarchical Clustering
+def busmap_by_hac(network, n_clusters, buses_i=None, branch_components=["Line", "Link"], feature=None):
+    """
+    Create a busmap accroding to Hierarchical Agglomerative Clustering.
+
+    Parameters
+    ----------
+    network : pypsa.Network
+    n_clusters : int
+        Final number of clusters desired.
+    buses_i: None|pandas.Index
+        Subset of buses to cluster. If None, all buses are considered.
+    branch_components: List
+        Subset of all branch_components in the network. Defaults to ["Line", "Link"].
+    feature: None | pandas.DataFrame
+        Feature to be considered for the clustering.
+        The DataFrame must be indexed with buses_i.
+        If None, all buses have the same similarity.
+
+    Returns
+    -------
+    busmap : pandas.Series
+        Mapping of network.buses to k-means clusters (indexed by
+        non-negative integers).
+    """
+
+    if find_spec('sklearn') is None:
+        raise ModuleNotFoundError("Optional dependency 'sklearn' not found."
+            "Install via 'conda install -c conda-forge scikit-learn' "
+            "or 'pip install scikit-learn'")
+
+    from sklearn.cluster import AgglomerativeClustering as HAC
+
+    if buses_i is None:
+        buses_i = network.buses.index
+
+    if feature is None:
+        logger.warning(
+            "No feature is specified for Hierarchical Clustering. "
+            "Falling back to default, where all buses have equal similarity. "
+            "You can specify a feature as pandas.DataFrame indexed with buses_i."
+        )
+
+        feature = pd.DataFrame(index=buses_i, columns=[""], data=0)
```

Comment:
```suggestion
            "You can specify a feature as pandas.Series indexed with buses_i."
        )

        feature = pd.Series(0, index=buses_i)
```

Are multiple features allowed, i.e. can the dataframe have multiple columns? If not, a pd.Series would be more elegant.

Reply:
would be, but does not work... and multiple columns are allowed. Throws the following error:

```
Expected 2D array, got 1D array instead:
array=[0 0 0 ... 0 0 0].
Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample.
```